extensions/common/iam/identity-trust/identity-trust-sts-embedded/src/main/java/org/eclipse/edc/iam/identitytrust/sts/embedded/EmbeddedSecureTokenService.java

```diff
@@ -0,0 +1,82 @@
+
+import org.eclipse.edc.identitytrust.SecureTokenService;
+import org.eclipse.edc.jwt.spi.JwtDecorator;
+import org.eclipse.edc.jwt.spi.TokenGenerationService;
+import org.eclipse.edc.spi.iam.TokenRepresentation;
+import org.eclipse.edc.spi.result.Result;
+import org.jetbrains.annotations.Nullable;
+
+import java.time.Clock;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Consumer;
+
+/**
+ * Implementation of a {@link SecureTokenService}, that is capable of creating a self-signed ID token ("SI token") completely in-process.
+ * To that end, it makes use of the <a href="https://connect2id.com/products/nimbus-jose-jwt">Nimbus JOSE/JWT library</a>.<br/>
+ * As a recommendation, the private key it uses should not be used for anything else.
+ */
+public class EmbeddedSecureTokenService implements SecureTokenService {
+
+    private final TokenGenerationService tokenGenerationService;
+    private final Clock clock;
+    private final long validity;
+
+
+    public EmbeddedSecureTokenService(TokenGenerationService tokenGenerationService, Clock clock, long validity) {
+        this.tokenGenerationService = tokenGenerationService;
+        this.clock = clock;
+        this.validity = validity;
+    }
+
+    @Override
+    public Result<TokenRepresentation> createToken(Map<String, Object> claims, @Nullable String bearerAccessScope) {
+        Collection<JwtDecorator> decorators = new ArrayList<>();
+        decorators.add(new SelfIssuedTokenDecorator(claims, clock, validity));
+
+        var result = withAccessToken(claims, bearerAccessScope, decorators::add);
+
+        if (result.failed()) {
+            return result.mapTo();
+        }
+        return tokenGenerationService.generate(decorators.toArray(JwtDecorator[]::new));
+    }
+
+    private Result<Void> withAccessToken(Map<String, Object> claims, String bearerAccessScope, Consumer<AccessTokenDecorator> consumer) {
+        return Optional.ofNullable(bearerAccessScope)
+                .map(scopes -> createAndAcceptAccessToken(claims, bearerAccessScope, consumer))
+                .orElse(Result.success());
+    }
+
+    private Result<Void> createAndAcceptAccessToken(Map<String, Object> claims, String bearerAccessScope, Consumer<AccessTokenDecorator> consumer) {
+        return createAccessToken(claims, bearerAccessScope)
+                .map(AccessTokenDecorator::new)
+                .onSuccess(consumer)
+                .map(deco -> null);
+    }
+
+    private Result<TokenRepresentation> createAccessToken(Map<String, Object> claims, String bearerAccessScope) {
+        return tokenGenerationService.generate(new SelfIssuedTokenDecorator(accessTokenClaims(claims), clock, validity), new BearerAccessScopeTokenDecorator(bearerAccessScope));
+    }
+
+    private Map<String, Object> accessTokenClaims(Map<String, Object> claims) {
+        return claims;
+    }
```

Comment:
Why this method as it only returns the input?

Reply:
Because the PR still in draft, i needed feedback from the CI :)

Reply:
done